packages/explorer/src/app/(explorer)/[chainName]/worlds/[worldAddress]/observe/TransactionsTableContainer.tsx

```diff
@@ -0,0 +1,197 @@
+  const { worldAddress } = useParams();
+  const wagmiConfig = useConfig();
+  const { data: worldAbiData } = useWorldAbiQuery();
+  const abi = worldAbiData?.abi;
+  const [transactions, setTransactions] = useState<WatchedTransaction[]>([]);
+  const observerWrites = useStore(store, (state) => Object.values(state.writes));
+
+  const mergedTransactions = useMemo((): WatchedTransaction[] => {
+    const mergedMap = new Map<Hex | undefined, WatchedTransaction>();
+
+    for (const write of observerWrites) {
+      const parsedAbiItem = parseAbiItem(`function ${write.functionSignature}`) as AbiFunction;
+      const functionData = {
+        functionName: parsedAbiItem.name,
+        args: write.args,
+      };
+
+      mergedMap.set(write.hash, {
+        status: "pending",
+        timestamp: BigInt(write.time) / 1000n,
+        functionData,
+        write,
+      });
+    }
+
+    for (const transaction of transactions) {
+      const existing = mergedMap.get(transaction.hash);
+      if (existing) {
+        mergedMap.set(transaction.hash, { ...transaction, write: existing.write });
+      } else {
+        mergedMap.set(transaction.hash, { ...transaction });
+      }
+    }
+
+    return Array.from(mergedMap.values()).sort((a, b) => Number(b.timestamp ?? 0n) - Number(a.timestamp ?? 0n));
+  }, [transactions, observerWrites]);
+
+  const handleTransaction = useCallback(
+    async (hash: Hex, timestamp: bigint) => {
+      if (!abi) return;
+
+      const transaction = await getTransaction(wagmiConfig, { hash });
+      if (transaction.to !== worldAddress) return;
+
+      let functionName: string | undefined;
+      let args: readonly unknown[] | undefined;
+      let transactionError: BaseError | undefined;
+
+      try {
+        const functionData = decodeFunctionData({ abi, data: transaction.input });
+        functionName = functionData.functionName;
+        args = functionData.args;
+      } catch (error) {
+        transactionError = error as BaseError;
+        functionName = transaction.input.length > 10 ? transaction.input.slice(0, 10) : "unknown";
+      }
+
+      setTransactions((prevTransactions) => [
+        {
+          hash,
+          timestamp,
+          transaction,
+          status: "pending",
+          functionData: {
+            functionName,
+            args,
+          },
+        },
+        ...prevTransactions,
+      ]);
+
+      // Receipt
+      const maxRetries = 10;
+      const retryInterval = 2000;
+      let receipt: TransactionReceipt | undefined;
+      for (let attempt = 0; attempt < maxRetries; attempt++) {
+        try {
+          receipt = await getTransactionReceipt(wagmiConfig, { hash });
+          if (receipt) break;
+        } catch (error) {
+          console.error(`Failed to fetch receipt, attempt ${attempt + 1}:`, error);
+        }
+
+        if (attempt < maxRetries - 1) {
+          await new Promise((resolve) => setTimeout(resolve, retryInterval));
+        }
+      }
```

Comment:
this is basically what `waitForTransactionReceipt` does, but with caching/batching - can we use that here?

Reply:
oh so nice!!! great suggestion, changed